### ISSUE TITLE
peel_to_type or generic peel methods on references

### ISSUE THREAD

**Issue by joshtriplett** [rust-lang/git2-rs#130]
Given a reference `r`, I tried to write the following:

``` rust
let c = r.peel(git2::ObjectType::Commit).unwrap_or_else(die).into_commit().unwrap();
```

(`unwrap_or_else(die)` here produces a clean error message and exits, unlike the assertion-failure-style error from `unwrap`.)

However, the Result returned by `into_commit` has the original `Object` as its error type, which doesn't support the `Debug` trait required for `unwrap`.  (And even if it did, an object isn't a useful error message.)

I understand the intent of `into_commit` and similar: they preserve either the object or its converted version.  However, in a context like the above (after calling `peel`) where it would be an internal error to have anything other than a given type of object, I'd like to have a convenient way to obtain the right type for the object.

To simplify this case, perhaps references could provide type-specific methods like `peel_to_commit`, `peel_to_tag`, and similar?  Or a method that has a generic return type and returns the object type the caller expects?

**Comment by alexcrichton** [comment]
Yeah I'd be pretty ok with something like `peel_to` perhaps? Wanna make a PR for it?

**Comment by vcfxb** [comment]
This can be closed as of ddf6f17b3a9dd986a6b311fc2c87140f57326ece perhaps? I'm looking through issues sorted by oldest and seeing what's outdated/closeable
